docs/proposals/20-ha-control-planes.md

```diff
@@ -0,0 +1,1259 @@
+
+The above list is not an exhaustive list and is just indicative that the currently set limit of 250 will have to be revisited.
+
+## Scheduling control plane components
+
+### Zone pinning
+
+HA shoot clusters with failure tolerance of `node` as well as non-HA shoot clusters can be scheduled on `single-zonal` and `multi-zonal` seeds alike. On a `multi-zonal` seed it's desireable to place components of the same control plane in **one zone only** to reduce cost and latency effects due to cross network traffic.
+Thus, it's essential to add [Pod affinity](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#inter-pod-affinity-and-anti-affinity) rules to control plane component with multiple replicas:
+
+```yaml
+spec:
+  affinity:
+    podAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+      - labelSelector:
+          matchLabels:
+            gardener.cloud/shoot: <technical-id>
+            <labels>
+        topologyKey: "topology.kubernetes.io/zone"
+```
+
+A special challenge is to select the entire set of control plane pods belonging to a single control plane. Today, Gardener and extensions don't put a common label to the affected pods. We propose to introduce a new label `gardener.cloud/shoot: <technical-id>` where `technical-id` is the shoot namespace. A mutating webhoook in the __Gardener Resource Manager__ should apply this label and the affinity rules to every pod in the control plane. This label and the pod affinity rule will ensure that all the pods in the control plane are pinned to a specific zone for HA shoot cluster having failure tolerance of `node`.
+
+### Single-Zone
+
+There are control plane components (like etcd) which requires one etcd member pod per node. Following anti-affinity rule guarantees that each pod of etcd is scheduled onto a different node.
+```yaml
+spec:
+  affinity:
+    podAntiAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+      - labelSelector:
+          matchLabels:
+            <labels>
+        topologyKey: "kubernetes.io/hostname"
+```
+For other control plane components which do not have a stricter requirements to have one replica per node, a more optimal scheduling strategy should be used. Following topology spread constraint provides better utilization of node resources, allowing cluster autoscaler to downsize node groups if certain nodes are under-utilized.
+
+```yaml
+spec:
+  topologySpreadConstraints:
+  - maxSkew: 1
+     topologyKey: kubernetes.io/hostname
+     whenUnsatisfiable: DoNotSchedule
+     labelSelector:
+      matchLabels:
+        <labels>
+```
+Using topology spread constraints (as described above) would still ensure that if there are more than one replica defined for a control plane component then it will be distributed across more than one node ensuring failure tolerance of at least one node.
+
+### Multi-Zone (#replicas == #zones)
```

Comment:
I guess this is also the approach to be taken for `#replicas < #zones`, which is a realistic scenario as there are regions with more than 3 zones. For completeness sake, I would add this to the document.

Reply:
yes that is correct. Made the change